fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/impl/transfer/transaction/TransactionImpl.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.transfer.transaction;
+
+import java.util.ArrayList;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.concurrent.locks.ReentrantLock;
+
+import net.fabricmc.fabric.api.transfer.v1.transaction.Participant;
+import net.fabricmc.fabric.api.transfer.v1.transaction.Transaction;
+import net.fabricmc.fabric.api.transfer.v1.transaction.TransactionResult;
+
+public class TransactionImpl implements Transaction {
+	private static Thread serverThread = null;
+	// outer lock: non-server threads contend on this one first.
+	private static final ReentrantLock outerLock = new ReentrantLock();
+	// inner lock: the server thread directly contends on this one with the non-server thread who owns the outerLock.
+	private static final ReentrantLock innerLock = new ReentrantLock();
+
+	private static final ArrayList<TransactionImpl> STACK = new ArrayList<>();
+	private static int stackPointer = -1;
+	// flag to prevent transaction functions from being called from within a transaction function.
+	// for example, to make sure that transaction functions are not called from Participant.onEnlist().
+	private static boolean allowAccess = true;
+
+	static {
+		STACK.add(new TransactionImpl());
+	}
+
+	@SuppressWarnings("rawtypes")
+	private final IdentityHashMap<Participant, Object> stateStorage = new IdentityHashMap<>();
+	private boolean isOpen = false;
+
+	public static void setServerThread(Thread serverThread) {
+		if (innerLock.isLocked()) {
+			throw new AssertionError("Something is terribly wrong.");
+		}
+
+		TransactionImpl.serverThread = serverThread;
+	}
+
+	// validate that Transaction instance methods are valid to call.
+	private void validateCurrent() {
+		if (!innerLock.isHeldByCurrentThread()) {
+			throw new IllegalStateException("Transaction operations are not allowed on this thread.");
```

Comment:
Maybe print the name of the thread you should be dispatching on?

Reply:
I don't think that information is available without subclassing `ReentrantLock`: there is a `getOwner` method, but it's protected.